src/nl/hannahsten/texifyidea/util/CommandRunner.kt

```diff
@@ -0,0 +1,127 @@
+package nl.hannahsten.texifyidea.util
+
+import com.intellij.execution.ExecutionException
+import com.intellij.execution.configurations.GeneralCommandLine
+import com.intellij.util.io.awaitExit
+import kotlinx.coroutines.*
+import java.io.File
+import java.io.IOException
+import javax.swing.SwingUtilities
+
+data class CommandResult(
+    val exitCode: Int,
+    val standardOutput: String?,
+    val errorOutput: String?
+) {
+
+    val output: String?
+        get() = if (standardOutput != null || errorOutput != null) standardOutput + errorOutput else null
+}
+
+/**
+ * Run a command in the terminal in a non-blocking way.
+ *
+ * @param workingDirectory If provided, the process' working directory.
+ * @param input If provided, this will be written to the process' input pipe.
+ * @param discardOutput Whether to discard all command outputs (stdout, stderr) and only return its exit code.
+ * @param returnExceptionMessageAsErrorOutput Whether to return exception messages as error output if exceptions are thrown.
+ * @param timeout The timeout for execution. Does not stop reading the process' output as long as it is available.
+ */
+suspend fun runCommandNonBlocking(
+    vararg commands: String,
+    workingDirectory: File? = null,
+    input: String? = null,
+    discardOutput: Boolean = false,
+    returnExceptionMessageAsErrorOutput: Boolean = false,
+    timeout: Long = 3
+): CommandResult = withContext(Dispatchers.IO) {
+    try {
+        Log.debug("isEDT=${SwingUtilities.isEventDispatchThread()} Executing in ${workingDirectory ?: "current working directory"} ${GeneralCommandLine(*commands).commandLineString}")
+
+        val processBuilder = GeneralCommandLine(*commands)
+            .withParentEnvironmentType(GeneralCommandLine.ParentEnvironmentType.CONSOLE)
+            .withWorkDirectory(workingDirectory)
+            .toProcessBuilder()
+
+        if (discardOutput) {
+            processBuilder.redirectOutput(ProcessBuilder.Redirect.DISCARD)
+            processBuilder.redirectError(ProcessBuilder.Redirect.DISCARD)
+        }
+
+        val process = processBuilder.start()
+
+        process.outputWriter().use { if (input != null) it.write(input) }
+        val output = if (!discardOutput) async { process.inputReader().use { it.readText() } } else null
+        val error = if (!discardOutput) async { process.errorReader().use { it.readText() } } else null
+
+        withTimeoutOrNull(1_000 * timeout) {
+            process.awaitExit()
+        } ?: run {
+            process.destroy()
+            Log.debug("${commands.firstOrNull()} exited ${process.exitValue()} after timeout $timeout seconds")
+        }
+
+        val exitCode = process.awaitExit()
+        Log.debug("${commands.firstOrNull()} exited with $exitCode (${output?.await()?.take(100)}, ${error?.await()?.take(100)})")
```

Comment:
```suggestion
        Log.debug("${commands.firstOrNull()} exited with $exitCode ${output?.await()?.take(100)} ${error?.await()?.take(100)}")
```
The extra parenthesis and comma confused me at first because I thought they were part of the output, I would prefer without

Reply:
Parenthesis and comma are removed.